cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultStageConfigurer.java

```diff
@@ -63,8 +67,28 @@ public void setOutputSchema(@Nullable Schema outputSchema) {
     this.outputSchema = outputSchema;
   }
 
-  public void addInputSchema(@Nullable String stageName, @Nullable Schema inputSchema) {
-    inputSchemas.put(stageName, inputSchema);
+  public void addInputSchema(String stageType, @Nullable String inputStageName, @Nullable Schema inputSchema) {
+    // Do not allow null input schema for Joiner
+    if (stageType.equalsIgnoreCase(BatchJoiner.PLUGIN_TYPE) && inputSchema == null) {
+        throw new IllegalArgumentException(String.format("Joiner cannot have any null input schemas, but stage %s " +
+                                                           "outputs a null schema.", inputStageName));
+    }
+
+    // Do not allow more than one input schema for stages other than Joiner
+    if (!stageType.equalsIgnoreCase(BatchJoiner.PLUGIN_TYPE) && !hasSameSchema(inputSchema)) {
+      throw new IllegalArgumentException("Two different input schema were set for the stage " + this.stageName);
+    }
+
+    inputSchemas.put(inputStageName, inputSchema);
+  }
+
+  private boolean hasSameSchema(Schema inputSchema) {
+    for (Schema schema : inputSchemas.values()) {
```

Comment:
its enough to just check for one of the schema, say the first entry. as you add the schemas serially, previous ones would have been checked already for equality..